File: packages/core/src/logs/logParser.ts
import { classify, stripTimestamp } from "./errorDetector";
import type { ClassifiedLine, ErrorBlock, ParsedLog } from "./types";

export interface ParseOptions {
  /** Truncate the log past this many lines (kept from the *end*). */
  maxLines?: number;
  /** Lines of context to include on either side of an error block. */
  contextLines?: number;
}

export class LogParser {
  private readonly maxLines: number;
  private readonly contextLines: number;

  constructor(opts: ParseOptions = {}) {
    this.maxLines = opts.maxLines ?? 5000;
    this.contextLines = opts.contextLines ?? 3;
  }

  parse(raw: string): ParsedLog {
    const allLines = raw.split(/\r?\n/);
    const truncated = allLines.length > this.maxLines
      ? allLines.slice(allLines.length - this.maxLines)
      : allLines;

    const lines: ClassifiedLine[] = truncated.map((rawLine, idx) => {
      const { severity, matchedPattern } = classify(rawLine);
      const line: ClassifiedLine = {
        number: idx + 1,
        text: stripTimestamp(rawLine),
        raw: rawLine,
        severity,
      };
      if (matchedPattern !== undefined) line.matchedPattern = matchedPattern;
      return line;
    });

    const errorLineNumbers = lines.filter((l) => l.severity === "error").map((l) => l.number);
    const warningLineNumbers = lines.filter((l) => l.severity === "warning").map((l) => l.number);

    const errorBlocks = groupBlocks(errorLineNumbers, lines, this.contextLines);
    const firstErrorLine = errorLineNumbers[0] ?? null;
    const summary = buildSummary({ totalLines: lines.length, errors: errorLineNumbers.length, warnings: warningLineNumbers.length });

    return { lines, errorLineNumbers, warningLineNumbers, errorBlocks, firstErrorLine, summary };
  }
}

function groupBlocks(errorLines: number[], lines: ClassifiedLine[], contextLines: number): ErrorBlock[] {
  if (errorLines.length === 0) return [];
  const blocks: ErrorBlock[] = [];
  let current: number[] = [];

  const flush = (): void => {
    if (current.length === 0) return;
    const start = Math.max(1, current[0]! - contextLines);
    const end = Math.min(lines.length, current[current.length - 1]! + contextLines);
    const lineNumbers: number[] = [];
    for (let n = start; n <= end; n++) lineNumbers.push(n);
    const preview = lineNumbers
      .map((n) => lines[n - 1]?.text ?? "")
      .join("\n");
    blocks.push({ lineNumbers, startLine: start, endLine: end, preview });
    current = [];
  };

  for (const lineNo of errorLines) {
    if (current.length === 0 || lineNo - current[current.length - 1]! <= contextLines + 1) {
      current.push(lineNo);
    } else {
      flush();
      current.push(lineNo);
    }
  }
  flush();
  return blocks;
}

function buildSummary(input: { totalLines: number; errors: number; warnings: number }): string {
  if (input.errors === 0 && input.warnings === 0) return `${input.totalLines} lines, no errors detected.`;
  const parts: string[] = [`${input.totalLines} lines`];
  if (input.errors > 0) parts.push(`${input.errors} error${input.errors === 1 ? "" : "s"}`);
  if (input.warnings > 0) parts.push(`${input.warnings} warning${input.warnings === 1 ? "" : "s"}`);
  return parts.join(", ") + ".";
}

/** Convenience: extract the first error block as plain text for "Copy Error Snippet". */
export function extractErrorSnippet(parsed: ParsedLog): string {
  const first = parsed.errorBlocks[0];
  if (!first) return parsed.summary;
  return first.preview;
}

File: packages/jetbrains-plugin/src/main/kotlin/com/actionlens/logs/LogParser.kt
package com.actionlens.logs

data class ClassifiedLine(
    val number: Int,
    val text: String,
    val raw: String,
    val severity: LineSeverity,
    val matchedPattern: String?,
)

data class ErrorBlock(
    val lineNumbers: List<Int>,
    val startLine: Int,
    val endLine: Int,
    val preview: String,
)

data class ParsedLog(
    val lines: List<ClassifiedLine>,
    val errorLineNumbers: List<Int>,
    val warningLineNumbers: List<Int>,
    val errorBlocks: List<ErrorBlock>,
    val firstErrorLine: Int?,
    val summary: String,
)

class LogParser(
    private val maxLines: Int = 5000,
    private val contextLines: Int = 3,
) {

    fun parse(raw: String): ParsedLog {
        val all = raw.split(Regex("\r?\n"))
        val truncated = if (all.size > maxLines) all.subList(all.size - maxLines, all.size) else all

        val lines = truncated.mapIndexed { idx, rawLine ->
            val (sev, pid) = ErrorDetector.classify(rawLine)
            ClassifiedLine(
                number = idx + 1,
                text = ErrorDetector.stripTimestamp(rawLine),
                raw = rawLine,
                severity = sev,
                matchedPattern = pid,
            )
        }

        val errorLineNumbers = lines.filter { it.severity == LineSeverity.ERROR }.map { it.number }
        val warningLineNumbers = lines.filter { it.severity == LineSeverity.WARNING }.map { it.number }
        val blocks = groupBlocks(errorLineNumbers, lines, contextLines)
        val firstError = errorLineNumbers.firstOrNull()
        val summary = buildSummary(lines.size, errorLineNumbers.size, warningLineNumbers.size)

        return ParsedLog(lines, errorLineNumbers, warningLineNumbers, blocks, firstError, summary)
    }

    private fun groupBlocks(errorLines: List<Int>, lines: List<ClassifiedLine>, ctx: Int): List<ErrorBlock> {
        if (errorLines.isEmpty()) return emptyList()
        val blocks = mutableListOf<ErrorBlock>()
        val current = mutableListOf<Int>()

        fun flush() {
            if (current.isEmpty()) return
            val start = maxOf(1, current.first() - ctx)
            val end = minOf(lines.size, current.last() + ctx)
            val nums = (start..end).toList()
            val preview = nums.joinToString("\n") { lines[it - 1].text }
            blocks += ErrorBlock(nums, start, end, preview)
            current.clear()
        }

        for (n in errorLines) {
            if (current.isEmpty() || n - current.last() <= ctx + 1) current += n
            else { flush(); current += n }
        }
        flush()
        return blocks
    }

    private fun buildSummary(total: Int, errors: Int, warnings: Int): String {
        if (errors == 0 && warnings == 0) return "$total lines, no errors detected."
        val parts = mutableListOf("$total lines")
        if (errors > 0) parts += "$errors error${if (errors == 1) "" else "s"}"
        if (warnings > 0) parts += "$warnings warning${if (warnings == 1) "" else "s"}"
        return parts.joinToString(", ") + "."
    }
}

object LogSanitiser {
    private val patterns = listOf<Pair<Regex, String>>(
        Regex("""\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{20,}""") to "[REDACTED:GITHUB_TOKEN]",
        Regex("""\bgithub_pat_[A-Za-z0-9_]{20,}""") to "[REDACTED:GITHUB_TOKEN]",
        Regex("""(?i)(authorization:\s*bearer\s+)[A-Za-z0-9._\-+/=]+""") to "$1[REDACTED]",
        Regex("""\bAKIA[0-9A-Z]{16}\b""") to "[REDACTED:AWS_KEY]",
        Regex("""(?i)\b(api[_-]?key|secret|password|token)\s*[:=]\s*['"]?[A-Za-z0-9._\-+/=]{8,}['"]?""") to "$1=[REDACTED]",
        Regex("""-----BEGIN [^-]+ PRIVATE KEY-----[\s\S]*?-----END [^-]+ PRIVATE KEY-----""") to "[REDACTED:PRIVATE_KEY]",
    )

    fun sanitise(text: String): String = patterns.fold(text) { acc, (r, repl) -> r.replace(acc, repl) }
}

data class AiDebugPromptInput(
    val repository: String,
    val branch: String? = null,
    val commitSha: String? = null,
    val workflowName: String? = null,
    val jobName: String? = null,
    val failedStepName: String? = null,
    val parsedLog: ParsedLog? = null,
    val snippet: String? = null,
)

private const val INSTRUCTION =
    "Analyze this GitHub Actions failure. Identify the root cause, explain why it happened, and suggest the minimal safe fix. Focus on the relevant log lines and avoid guessing."

object AiDebugPrompt {
    fun create(input: AiDebugPromptInput): String {
        val sb = StringBuilder()
        sb.appendLine(INSTRUCTION)
        sb.appendLine()
        sb.appendLine("Context:")
        sb.appendLine("- Repository: ${input.repository}")
        input.branch?.let { sb.appendLine("- Branch: $it") }
        input.commitSha?.let { sb.appendLine("- Commit: $it") }
        input.workflowName?.let { sb.appendLine("- Workflow: $it") }
        input.jobName?.let { sb.appendLine("- Job: $it") }
        input.failedStepName?.let { sb.appendLine("- Failed step: $it") }

        val snippet = input.parsedLog?.errorBlocks?.firstOrNull()?.preview ?: input.snippet
        if (!snippet.isNullOrBlank()) {
            sb.appendLine()
            sb.appendLine("Relevant log lines:")
            sb.appendLine("```")
            sb.appendLine(LogSanitiser.sanitise(snippet).trimEnd())
            sb.appendLine("```")
        }
        input.parsedLog?.let {
            sb.appendLine()
            sb.appendLine("Summary: ${it.summary}")
        }
        return sb.toString().trimEnd()
    }
}
